staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go

```diff
@@ -36,9 +36,28 @@ import (
 // timeout of the HTTP request, including reading the response body.
 const defaultRequestTimeout = 30 * time.Second
 
+// GenericWebhook defines a generic client for weboooks with commonly used capabilities,
+// such as retry requests.
 type GenericWebhook struct {
 	RestClient     *rest.RESTClient
 	InitialBackoff time.Duration
+	ShouldRetry    func(error) bool
+}
+
+// DefaultShouldRetry is a default implementation for the GenericWebhook ShouldRetry function property.
+// If the error reason is one of: networking (connection reset) or http (InternalServerError (500), GatewayTimeout (504), TooManyRequests (429)),
+// or apierrors.SuggestsClientDelay() returns true, then the function advises a retry.
+// Otherwise it returns false for an immediate fail.
+func DefaultShouldRetry(err error) bool {
+	// these errors indicate a transient error that should be retried.
+	if net.IsConnectionReset(err) || apierrors.IsInternalError(err) || apierrors.IsTimeout(err) || apierrors.IsTooManyRequests(err) {
+		return true
+	}
+	// if the error sends the Retry-After header, we respect it as an explicit confirmation we should retry.
+	if _, shouldRetry := apierrors.SuggestsClientDelay(err); shouldRetry {
```

Comment:
I realize this is pre-existing code. But it looks like we don't fully respect the header; we don't wait for however long Retry-After requests us to. Could we fix that now?

Reply:
I agree it's kind of "almost there" solution. But do we want to risk waiting as long as an external webhook tells us to, or we want the api server to have full control on the wait time between retries? And equally, does it make sense to retry before the recommended time at all? It seems to me that both approaches have downsides.

I would move the discussion in a separate PR though as this one is not precisely for this.

Reply:
Would you open up an issue for this?

Reply:
sure: https://github.com/kubernetes/kubernetes/issues/83377